cpp/include/raft/neighbors/detail/cagra/compute_distance.hpp

```diff
@@ -36,152 +36,54 @@ _RAFT_DEVICE constexpr unsigned get_vlen()
   return utils::size_of<LOAD_T>() / utils::size_of<DATA_T>();
 }
 
-template <class LOAD_T, class DATA_T, unsigned VLEN>
-struct data_load_t {
-  union {
-    LOAD_T load;
-    DATA_T data[VLEN];
-  };
+template <std::uint32_t VECLEN>
+struct code_book_load_t_core {
+  using type = void;
 };
-
-template <class LOAD_T,
-          class DATA_T,
-          class DISTANCE_T,
-          std::uint32_t DATASET_BLOCK_DIM,
-          std::uint32_t TEAM_SIZE,
-          bool use_reg_fragment>
-struct distance_op;
-template <class LOAD_T,
-          class DATA_T,
-          class DISTANCE_T,
-          std::uint32_t DATASET_BLOCK_DIM,
-          std::uint32_t TEAM_SIZE>
-struct distance_op<LOAD_T, DATA_T, DISTANCE_T, DATASET_BLOCK_DIM, TEAM_SIZE, false> {
-  const float* const query_buffer;
-  __device__ distance_op(const float* const query_buffer) : query_buffer(query_buffer) {}
-
-  __device__ DISTANCE_T operator()(const DATA_T* const dataset_ptr,
-                                   const std::uint32_t dataset_dim,
-                                   const bool valid)
-  {
-    const unsigned lane_id  = threadIdx.x % TEAM_SIZE;
-    constexpr unsigned vlen = get_vlen<LOAD_T, DATA_T>();
-    constexpr unsigned reg_nelem =
-      (DATASET_BLOCK_DIM + (TEAM_SIZE * vlen) - 1) / (TEAM_SIZE * vlen);
-    data_load_t<LOAD_T, DATA_T, vlen> dl_buff[reg_nelem];
-
-    DISTANCE_T norm2 = 0;
-    if (valid) {
-      for (uint32_t elem_offset = 0; elem_offset < dataset_dim; elem_offset += DATASET_BLOCK_DIM) {
-#pragma unroll
-        for (uint32_t e = 0; e < reg_nelem; e++) {
-          const uint32_t k = (lane_id + (TEAM_SIZE * e)) * vlen + elem_offset;
-          if (k >= dataset_dim) break;
-          dl_buff[e].load = *reinterpret_cast<const LOAD_T*>(dataset_ptr + k);
-        }
-#pragma unroll
-        for (uint32_t e = 0; e < reg_nelem; e++) {
-          const uint32_t k = (lane_id + (TEAM_SIZE * e)) * vlen + elem_offset;
-          if (k >= dataset_dim) break;
-#pragma unroll
-          for (uint32_t v = 0; v < vlen; v++) {
-            const uint32_t kv = k + v;
-            // if (kv >= dataset_dim) break;
-            DISTANCE_T diff = query_buffer[device::swizzling(kv)];
-            diff -= spatial::knn::detail::utils::mapping<float>{}(dl_buff[e].data[v]);
-            norm2 += diff * diff;
-          }
-        }
-      }
-    }
-    for (uint32_t offset = TEAM_SIZE / 2; offset > 0; offset >>= 1) {
-      norm2 += __shfl_xor_sync(0xffffffff, norm2, offset);
-    }
-    return norm2;
-  }
+template <>
+struct code_book_load_t_core<1> {
```

Comment:
Could we use `TxN_t` https://github.com/rapidsai/raft/blob/3e67a6c3bc5c73cd4a77a71bb5c2d911f8a7bcd4/cpp/include/raft/util/vectorized.cuh#L238-L249

Reply:
I changed to `TxN_t`